### PR TITLE
escape pg ? operator in proxy

### LINKED ISSUE #27892: [proxy] not treat "?" operator correctly in postgresql
### Which version of ShardingSphere did you use?
5.3.0

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
proxy 

### Expected behavior
pgsql  execute success 

 SELECT  * from table  WHERE receiver ?& $1 OR creater = $2;    
 parameter   1, test@gmail.com     2, susongyan

### Actual behavior
org.postgresql.util.PSQLException: No value specified for parameter 3.

### Reason analyze (If you can)
PostgreSQLComParseExecutor #convertSQLToJDBCStyle

after proxy parse this sql and convert it to jdbc paramter style 
the actual sql to execute in jdbc is    "SELECT  * from table  WHERE receiver ?& ? OR creater = ?;"

jdbc style, treat "?" as parameter marker, then  this sql  need three parameters in mistake (actually only two)

### PR DESCRIPTION
**Pull request by susongyan**
Fixes #27892.

Changes proposed in this pull request:
  -

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by azexcy** [comment]
https://github.com/apache/shardingsphere/actions/runs/5758958821/job/15612284497?pr=27909

**Comment by sandynz** [comment]
Hi @susongyan 
The pipeline E2E of PostgreSQL and openGauss failed, e.g. 
> error message is: The URL `jdbc:postgresql://postgresql_0.host:5432/pipeline_it_0??ssl=on&sslmode=prefer&stringtype=unspecified&TimeZone=UTC&bitToString=true` is not recognized, please refer to the pattern `(?<schema>[\w-.+:%%]+)\s*(?://(?<authority>[^/?#]*))?\s*(?:/(?!\s*/)(?<path>[^?#]*))?(?:\?(?!\s*\?)(?<query>[^#]*))?`.]

Looks `pipeline_it_0??ssl=on` is wrong, there're two `?`. Could you debug to verify the `sql` values?

You could refer to [Pipeline E2E test]( https://shardingsphere.apache.org/document/current/cn/test-manual/pipeline-e2e-test/ ) to run it on local.

**Comment by susongyan** [comment]
> 

ok,  let me check it

**Comment by susongyan** [comment]
@sandynz  plz take a look

**Comment by sandynz** [comment]
Hi @susongyan , some questions:

1, Why E2E failed before

2, I had a look at your supplied PostgreSQL jsonb doc. I'm afraid `sql.replace("?", "??")` might replace `?` in other cases we don't know for now

And I'll invite another member to review it.

**Comment by susongyan** [comment]
> 

to question 1:    i  dosen't  check is dml stamtent ,    failed  e2e case is   migration sql (distsql)  that contains jdbc connection string 

so now just deal  jsonb '?' operator in  dml sql

**Comment by sandynz** [comment]
OK. I just commented on issue about `Parser.parseJdbcSql`, still need to verify it.

File: proxy/frontend/type/postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/parse/PostgreSQLComParseExecutor.java [hunk 1]
```diff
@@ -87,6 +93,13 @@ private SQLParserEngine createShardingSphereSQLParserEngine(final String databas
         return sqlParserRule.getSQLParserEngine(protocolType.getTrunkDatabaseType().orElse(protocolType));
     }
     
+    private String escape(final SQLStatement sqlStatement, final String sql) {
+        if (sqlStatement instanceof DMLStatement) {
+            return sql.replace("?", "??");
+        }
+        return sql;
+    }
```
Comment (TeslaCN) [inline proxy/frontend/type/postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/parse/PostgreSQLComParseExecutor.java:101]: I found the doc about escape `?` in the following link. I think the escape logic could be merged into `convertSQLToJDBCStyle` and invoke the method with `convertSQLToJDBCStyle(Collections.emptyList(), sql)` if parameter placeholder is absent.
https://jdbc.postgresql.org/documentation/query/#using-the-statement-or-preparedstatement-interface

Comment (susongyan) [reply proxy/frontend/type/postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/parse/PostgreSQLComParseExecutor.java:101]: i found, if  this escape  moved to    convertSQLToJDBCStyle,   the sqlStatement's parameter mark indexes will be missed up

File: proxy/frontend/type/postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/parse/PostgreSQLComParseExecutor.java [hunk 2]
```diff
@@ -61,6 +62,11 @@ public Collection<DatabasePacket> execute() {
         SQLParserEngine sqlParserEngine = createShardingSphereSQLParserEngine(connectionSession.getDatabaseName());
         String sql = packet.getSQL();
         SQLStatement sqlStatement = sqlParserEngine.parse(sql, true);
+        String escapedSql = escape(sqlStatement, sql);
+        if (!escapedSql.equalsIgnoreCase(sql)) {
+            sqlStatement = sqlParserEngine.parse(escapedSql, true);
```
Comment (TeslaCN) [inline proxy/frontend/type/postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/parse/PostgreSQLComParseExecutor.java:67]: Is `??&` a valid PostgreSQL syntax? If we escape `?` here, maybe parsing SQL contains `??&` should be failed?

Comment (TeslaCN) [reply proxy/frontend/type/postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/command/query/extended/parse/PostgreSQLComParseExecutor.java:67]: Ignore it. Our PostgreSQL parser has already treated `?` as valid syntax.

**Review by TeslaCN — APPROVED** [review]

**Pull request merged**